apollo-backend/src/main/java/io/logz/apollo/kubernetes/KubernetesMonitor.java

```diff
@@ -97,34 +103,40 @@ public void stop() {
-                        }
-                        break;
-                    case PENDING_CANCELLATION:
-                        returnedDeployment = kubernetesHandler.cancelDeployment(deployment);
-                        break;
-                    default:
-                        returnedDeployment = kubernetesHandler.monitorDeployment(deployment);
-                        break;
-                }
-
-                deploymentDao.updateDeploymentStatus(deployment.getId(), returnedDeployment.getStatus());
-
-                if (deployment.getStatus().equals(Deployment.DeploymentStatus.DONE) || deployment.getStatus().equals(Deployment.DeploymentStatus.CANCELED)) {
-                    deploymentEnvStatusManager.updateDeploymentEnvStatus(deployment, deploymentEnvStatusManager.getDeploymentCurrentEnvStatus(deployment, kubernetesHandler));
+                if (!scopedEnvironments.contains(deployment.getEnvironmentId())) {
+                    logger.info("Deployment {} is of environment {} which is out of scope for me, skipping.",
+                            deployment.getId(), deployment.getEnvironmentId());
+                } else {
+                    Environment relatedEnv = environmentDao.getEnvironment(deployment.getEnvironmentId());
+                    KubernetesHandler kubernetesHandler = kubernetesHandlerStore.getOrCreateKubernetesHandler(relatedEnv);
+
+                    Deployment returnedDeployment;
+
+                    switch (deployment.getStatus()) {
+                        case PENDING:
+                            if (isDeployAllowed(deployment, environmentDao, deploymentDao)) {
+                                returnedDeployment = kubernetesHandler.startDeployment(deployment);
+                            } else {
+                                logger.info("Environment {} concurrency limit reached, not starting new deployment {} until one is done.", deployment.getEnvironmentId(), deployment.getId());
+                                returnedDeployment = deployment;
+                            }
+                            break;
+                        case PENDING_CANCELLATION:
+                            returnedDeployment = kubernetesHandler.cancelDeployment(deployment);
+                            break;
+                        default:
+                            returnedDeployment = kubernetesHandler.monitorDeployment(deployment);
+                            break;
+                    }
+
+                    deploymentDao.updateDeploymentStatus(deployment.getId(), returnedDeployment.getStatus());
+
+                    if (deployment.getStatus().equals(Deployment.DeploymentStatus.DONE) || deployment.getStatus().equals(Deployment.DeploymentStatus.CANCELED)) {
+                        deploymentEnvStatusManager.updateDeploymentEnvStatus(deployment, deploymentEnvStatusManager.getDeploymentCurrentEnvStatus(deployment, kubernetesHandler));
+                    }
```

Comment:
Extract to a different method

Reply:
Which part do you mean?

Reply:
The body of the else block of this condition: 
`if (!scopedEnvironments.contains(deployment.getEnvironmentId())) {`